src/include/daos_fs_sys.h

```diff
@@ -0,0 +1,581 @@
+					&dfs_sys->dfs_hash);
+		if (rc) {
+			/** TODO error handling */
+			D_GOTO(err_hash, rc);
+		}
+	}
+
+err_hash:
+	dfs_umount(dfs_sys->dfs);
+out:
+	return rc;
+}
+
+/**
+ * Unmount a file system with dfs_mount and destroy the hash.
+ */
+int
+dfs_sys_umount(dfs_sys_t *dfs_sys)
+{
+	int rc;
+
+	/** TODO check params to this function */
+	rc = d_hash_table_destroy(dfs_sys->dfs_hash, true);
+	/** TODO error handling */
+
+	rc = dfs_umount(dfs_sys->dfs);
+	/** TODO error handling */
+
+	return rc;
+}
+
+
+/**
+ * Try to get name from the hash.
+ * If not found, call dfs_lookup on name
+ * and store in the hash.
+ * Stores the hashed obj in _obj.
+ */
+static int
+dfs_sys_hash_lookup(const char *name, dfs_sys_t *dfs_sys, dfs_obj_t **_obj)
+{
+	struct dfs_sys_hash_hdl	*hdl;
+	d_list_t		*rlink;
+	size_t			name_len;
+	int			rc;
+
+	/** Make sure the hash is initialized */
+	/** TODO optionally allow no caching? */
+	if (dfs_sys->dfs_hash == NULL)
+		return EINVAL;
+
+	name_len = strnlen(name, PATH_MAX);
+	if (name_len > PATH_MAX-1)
+		return ENAMETOOLONG;
+
+	/** If cached, return it */
+	rlink = d_hash_rec_find(dfs_sys->dfs_hash, name, name_len);
+	if (rlink != NULL) {
+		hdl = dfs_sys_hash_hdl_obj(rlink);
+		D_GOTO(out, rc = 0);
+	}
+
+	/** Not cached, so add it */
+	hdl = dfs_sys_hash_hdl_new();
+	if (hdl == NULL)
+		return ENOMEM;
+
+	D_STRNDUP(hdl->name, name, PATH_MAX);
+	if (hdl->name == NULL)
+		return ENOMEM;
+
+	/** Lookup name in dfs */
+	rc = dfs_lookup(dfs_sys->dfs, name, O_RDWR, &hdl->oh, NULL, NULL);
+	if (rc) {
+		D_ERROR("dfs_lookup() %s failed (%d)\n", name, rc);
+		D_GOTO(err_hdl, rc);
+	}
+
+	/* Store this entry in the hash.
+	 * Since we have already called d_hash_rec_find,
+	 * pass exclusive=false to avoid another find being called */
+	rc = d_hash_rec_insert(dfs_sys->dfs_hash, hdl->name, name_len,
+			       &hdl->entry, false);
+	if (rc)
+		D_GOTO(err_hdl, rc = daos_der2errno(rc));
+
+out:
+	*_obj = hdl->oh;
+	return rc;
+err_hdl:
+	dfs_sys_hash_hdl_delete(hdl);
+	return rc;
+}
+
+static int parse_filename(const char* path, char** _obj_name, char** _cont_name)
+{
+	/**
+	 * TODO clean up this function for this context.
+	 * It was copied from somewhere else, so some parts may not be relevant.
+	 * Better error handling.
+	 * Double check for memory leaks.
+	 * Possible efficiency improvements?
+	 * Don't support "." and cwd? Since it doesn't make sense for a cont
+	 */
+	char	*f1 = NULL;
+	char	*f2 = NULL;
+	char	*fname = NULL;
+	char	*cont_name = NULL;
+	int	rc = 0;
+
+	if (path == NULL || _obj_name == NULL || _cont_name == NULL)
+		return -EINVAL;
+
+	if (strcmp(path, "/") == 0) {
+		*_cont_name = strdup("/");
+		if (*_cont_name == NULL)
+			return -ENOMEM;
+		*_obj_name = NULL;
+		return 0;
+	}
+
+	f1 = strdup(path);
+	if (f1 == NULL) {
+		rc = -ENOMEM;
+		goto out;
+	}
+
+	f2 = strdup(path);
+	if (f2 == NULL) {
+		rc = -ENOMEM;
+		goto out;
+	}
+
+	fname = basename(f1);
+	cont_name = dirname(f2);
```

Comment:
It is well worth reading the man page for these two functions and understanding all the implications of them.